core/src/code.rs

```diff
@@ -62,35 +87,71 @@ fn get_exports(
         .ok_or(CodeError::TypeSectionNotFound)?
         .types();
 
-    let import_count = module
+    let import_count = module.import_count(ImportCountType::Function);
+
+    let imports = module
         .import_section()
         .ok_or(CodeError::ImportSectionNotFound)?
-        .functions();
+        .entries();
 
-    for entry in module
+    let exports = module
         .export_section()
         .ok_or(CodeError::ExportSectionNotFound)?
-        .entries()
-        .iter()
-    {
-        if let Internal::Function(i) = entry.internal() {
-            if reject_unnecessary {
+        .entries();
+
+    if config.check_exports {
+        let mut entry = false;
+        for export in exports {
+            if let Internal::Function(i) = export.internal() {
                 // Index access into arrays cannot panic unless the Module structure is invalid
                 let type_id = funcs[*i as usize - import_count].type_ref();
                 let Type::Function(ref f) = types[type_id as usize];
                 if !f.params().is_empty() || !f.results().is_empty() {
                     return Err(CodeError::InvalidExportFnSignature);
                 }
+                if !ALLOWED_EXPORTS.contains(&export.field()) && config.check_exports {
+                    return Err(CodeError::NonGearExportFnFound);
+                }
+                if REQUIRED_EXPORTS.contains(&export.field()) {
+                    entry = true;
+                }
             }
-            if let Some(kind) = DispatchKind::try_from_entry(entry.field()) {
-                exports.insert(kind);
-            } else if !STATE_EXPORTS.contains(&entry.field()) && reject_unnecessary {
-                return Err(CodeError::NonGearExportFnFound);
+        }
+
+        if !entry {
+            return Err(CodeError::RequiredExportFnNotFound);
+        }
+    }
+
+    if config.check_imports {
+        for import in imports {
+            if let External::Function(i) = import.external() {
+                let Type::Function(types) = &types[*i as usize];
+                // We can likely improve this by adding some helper function in SyscallName
+                let syscall = SyscallName::all()
```

Comment:
1. use `SyscallName::instrumentable` instead of `SyscallName::all()` (because of `SystemBreak` syscall)
2. `instrumentable` will return `BTreeSet<_>` and you can do `syscalls_set.contains(syscall_name)` which is more efficient than iteration for each import
3. prohibit importing syscall more than once (this is what embedded_executor does when instantiating a module)

Reply:
One issue here is that we dont have a from_str on SyscallName, so I cannot just search the Set for it, instead having to iterate again and check which one matches the to_string() signature. Perhaps I could make some kind of Map<String, SyscallName> by iterating over the set once, then mapping the str to each one. That is what I will do for now, if there is a better solution, I will replace it with that.

Reply:
- plz remove `// We can likely improve this by adding some helper function in SyscallName`
- `from_str` doesn't seem to be appropriate because we only want to call it for instrumentable syscalls
- it would be nice to implement point 3 from previous comment